### ISSUE TITLE
Handle MountFlags for mount units

### ISSUE THREAD

**Issue by g0tar** [systemd/systemd#21873]
There was a discussion about changing Linux's default `private` propagation mode to `shared` in systemd. While there are valid points in both approaches and choosing The Right One seems a matter of taste, there should be a way to override the option for specific case. Mine being a normal, manual mounts (so not a subject for remount --make-priivate or some C coding) and inconveniences that might arise during regular system maintenance. Consider a situation:

```
S: unshare -m --propagation unchanged zsh
M: mkdir /dev/shm/temp
M: mount -t tmpfs test-mount /dev/shm/temp
S: cd /dev/shm/temp
M: umount /dev/shm/temp
umount: /dev/shm/temp: target is busy.
```

My guess is that it would be the most straightforward and expected to allow sysadmin to create some drop-in (or append `x-systemd.mountflags=private`) to configure appropriately his mountpoints, according to actually foreseen use-cases (considering general mounts that should, and the private mounts that shall not propagate), than falling back to  "what they **usually** want" wild-guesses and hand-crafting units doing:

> Setups which prefer the default of "private" should undo this change via invoking "mount --make-private /" or a similar command after boot.

I would prefer to mount all the virtual filesystems with `private` (as the default rule, since my / resides on real fs), which would also remedy the "umount shared /dev/pts blowing up the system" problems reported back then in some discussions about changed default.

The case above can be solved in different ways, but all of them require throwing a bunch of additional commands, while /dev being mounted as `shared` seems simply ...pointless. Those additional commands are nothing more than PITA and the problem grows with number of mountpoints, so this becomes irritating very fast especially on servers:

```
$ lsblk | wc -l
69
```

In short: on all my servers I don't need **any** temporary mounts nor removable media to be visible in **any** namespace. On the contrary, **all** the fs-namespaces that will be spawned should see nothing except their **own** mountpoints, so they should not receive anything from host, but more on this in separate FR as this seems currently impossible with systemd.

**Comment by poettering** [comment]
> S: unshare -m --propagation unchanged zsh
> M: mkdir /dev/shm/temp
> M: mount -t tmpfs test-mount /dev/shm/temp
> S: cd /dev/shm/temp
> M: umount /dev/shm/temp
> umount: /dev/shm/temp: target is busy.

No sure I follow? Our default propagation on the host is shared. Since you didn't change the propagation of the mounts the umount should just propagate. back to the main mount namespace. What am I missing?

**Comment by poettering** [comment]
> The case above can be solved in different ways, but all of them require throwing a bunch of additional commands, while /dev being mounted as `shared` seems simply ...pointless.

Not sure I follow? Why pointless? If you run a system service with PrivatTmp=yes and no other service settings the effect should be that /tmp and /var/tmp for that service get replaced by something service specific but /dev should be shraed from the host. We can only implement that if /dev/ is mounted as shared on the host, because otherwise propagation will be off...

What am I missing? Not following at all what you are trying to say?

**Comment by g0tar** [comment]
> > S: cd /dev/shm/temp
> > M: umount /dev/shm/temp
> > umount: /dev/shm/temp: target is busy.
> 
> No sure I follow? Our default propagation on the host is shared. Since you didn't change the propagation of the mounts the umount should just propagate. back to the main mount namespace. What am I missing?

I'm umounting **from** the main mountns and it does try to propagate to the slave, that works as expected and documented (this is RFE not bug report). The problem is, slave keeps the mountpoint busy (`cd` above). This is artificial example of course but shows what might happen in real life (e.g. remounting `/dev/shm` vs some namespaced service starting using shm). My point is - there are **some** consequences from shared vs private and I'm not sure whether they were all considered (and the risks simply accepted), but anyway these consequences should be resolved by system admin, i.e. be **configurable**, unless the defaults chosen are collision-free (therefore more fine-tuned I think).

In other words: is there any point for `/dev/shm` (in particular) to be mounted with shared propagation? Is that rationale valid for other (all) virtual fses? If there is one exception, this could be hardcoded, but with more exceptions, it should be configurable.

More than above, the forced (like: not configurable and global) shared propagation seems to me a bit off from the most powerful features of `systemd` being the resources and scope isolation, as shared `/dev/shm` is additional exposition vector that must be  handled manually somehow.

Currently, when one wants some filesystem (like `/dev/shm` in my case) to have private propagation, he must do some hand-crafting with `mount --make-private ...` as propagation model is not `fstab`-able. This ticket is about enhancement allowing changing default (per system) propagation model of specific mountpoint, so the admin doesn't have to remember doing it manually every time.

**Comment by poettering** [comment]
> n other words: is there any point for `/dev/shm` (in particular) to be mounted with shared propagation? Is that rationale valid for other (all) virtual fses? If there is one exception, this could be hardcoded, but with more exceptions, it should be configurable.

We generally follow the model that you can turn on knobs like PrivateTmp= or so that give you the same environment that you would usually get — except that you get your own /tmp and /var/tmp. The other PrivateXYZ=/ProtectXYZ= settings work the same way: most is passed thru as is, but some bits obstructred/replaced. Now, if /dev/shm would be mounted private in the host mount namespace, then turning any of those options on would implicitly mean /dev/shm would *disappear* entirely for those serviecs, there would be the empty underlying dir inode there. And that's not desirable at all. It just means shared memory obejcts would be placed in /dev instead, which is really bad idea, and half of the services would see them and others would not.

If you want to isolate /dev/shm then you need to use TemporaryFileSystem=/dev/shm or so, because then you not only want to make the old mount go away, but add a *new* instance there in the environment.

So, no "private" i the host namespace is typically not what we want, since via the knobs we expose we do denylisting and not allowlisting of file system objects and even if we didn't want to propagate some mounts we even less want the underlying inodes to just show up in the mount namespace...

**Comment by poettering** [comment]
Anyway, I am not opposed to making this configurable per mount point, but I am very sure the default should be "shared". And if you want to part from that, do it locally, but I have the suspicion there's some confusion on your side left what this actually means.

More importantly though: I am pretty sure this should be implemented in /bin/mount itself, i.e. add a new X-mount.private mount option or so that is processed by /bin/mount right after establishing the mount. i.e. similar to X-mount.mkdir.

Hence, my suggestion would be: please file an RFE issue on util-linux instead. i.e. this is othing tht should work in .mount only, it should also work in manual command lines and fstab and so on.

/cc @karelzak

**Comment by g0tar** [comment]
> Not sure I follow? Why pointless? If you run a system service with PrivatTmp=yes and no other service settings the effect should be that /tmp and /var/tmp for that service get replaced by something service specific but /dev should be shraed from the host. We can only implement that if /dev/ is mounted as shared on the host, because otherwise propagation will be off...

OK, I think I get it. The model assumes using specific `Private*` for explicit isolation (blacklisting), so the fallback must be "share"; this seems intricate, but valid. Combined with `MountImages` from referenced issue, although obscure, covers most of the usecases I can think of. Except the one, so allow me to follow here.

While the `shared` as default seems rational (for removable media, system mountpoints, apparently for virtual filesystems as well) due to the various `Private*` flags and containers spawning, this still ignores the fact, that there might be some fixed mountpoints that the system admin **knows** that are not subject to be propagated anywhere. Currently there's no way to perform such mount easily in race-free manner, as the remount must be done from some other unit. This could also be resolved by handling some `x-systemd.mountflags=private` from fstab (or `MountFlags=private` in mount unit).

**Comment by poettering** [comment]
Note that doing it entirely racefree is hard, since the traditional mount api of Linux at least doesn't allow the propagation settings to be set atomically with the creation of the mount point. But the new mount API might. But I think this is something /bin/mount should abstract away.

**Comment by karelzak** [comment]
You can use propagation flags in /etc/fstab, man mount(8):
```
       Since util-linux 2.23 the mount command can be used to do more propagation (topology)  changes  by
       one mount(8) call and do it also together with other mount operations.  This feature is EXPERIMEN‐
       TAL.  The propagation flags are applied by additional mount(2) system  calls  when  the  preceding
       mount operations were successful.  Note that this use case is not atomic.  It is possible to spec‐
       ify the propagation flags in fstab(5) as mount options (private, slave, shared, unbindable,  rpri‐
       vate, rslave, rshared, runbindable).
```

And I guess it will be atomic in near future after upgrading libmount to new kernel mount APIs.

**Comment by poettering** [comment]
Ah, excellent. I thought I had seen that but couldn't find this in the man page. OK, let's close this issue then, as this apparently is already available.

**Comment by g0tar** [comment]
Thanks guys for clarifications, I really appreciate your help here! It crossed my mind that this might belong to `mount` ...and there it is, I totally missed it. Everything works like expected

BTW I found the `/boot` mountpoint as a generic candidate for `private` (reducing exposure of kernel image and `System.map` for the same reasons like `kallsyms` via `kptr_restrict`); as EFI ESP partitions are FAT-formatted and lack the fine-grained DAC bits it's very easy to miss securing it properly; with `private` the attack surface could be significantly smaller.
